src/compiler/parser.ts

```diff
@@ -3981,19 +3981,31 @@ module ts {
                     return parseDebuggerStatement();
                 case SyntaxKind.AtToken:
                     return parseDeclaration();
-                case SyntaxKind.ConstKeyword:
                 case SyntaxKind.DeclareKeyword:
+                    if (lookAhead(isFollowedByIdentifierOrKeywordOnSameLine) && getDeclarationFlags() & flags) {

```

Comment:
You need some comment explaining why the lookahead is necessary here, but not for the syntaxkind kinds below.